### PR TITLE
Fuse LayerNormalization for Apex O2

### PR DESCRIPTION
**Pull request by iK1D**
When using Apex O2, a Cast node will be added between Div and Mul in the LayerNorm sub-graph. The input has float type and the weight and bias is float16 type. Change the LayerNorm related Ops to support input and weight/bias having different type, and change LayerNorm fusion to support the new pattern. This is helpful for models using Apex O2 to train. For some transformer models, before the change, ORT perf is similar to PyTorch, after the change, ORT will have >15% perf gain.

### CODE REVIEW

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 1]
```diff
@@ -1982,22 +1982,25 @@ void RegisterContribSchemas() {
             AttributeProto::INT, static_cast<int64_t>(ONNX_NAMESPACE::TensorProto_DataType_FLOAT))
       .AllowUncheckedAttributes()
       .Input(0, "X", "Input data tensor from the previous layer.", "T")
-      .Input(1, "Scale", "Scale tensor.", "T")
-      .Input(2, "B", "Bias tensor.", "T", OpSchema::Optional)
-      .Output(0, "Y", "Output data tensor.", "T")
+      .Input(1, "Scale", "Scale tensor.", "V")
+      .Input(2, "B", "Bias tensor.", "V", OpSchema::Optional)
+      .Output(0, "Y", "Output data tensor.", "V")
```
Comment (pengwa) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1987]: >V [](http://example.com/codeflow?start=46&length=1)

Inputs of type (T, V) will produce type (V and U). 
So for cases T=float, V=half.
We would allow Input <T=float, V=half> --> Output <V=half, U=float>. 

One of the problem is, the intermediate results ideally need a safe data type (after input type promotion) , do we have some logics  gurantee this in the kernel implementation?

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 2]
```diff
@@ -2092,9 +2096,9 @@ void RegisterContribSchemas() {
                 .Add("StdDev = Sqrt (VarPlusEpsilon)")
                 .Add("Deviation = Sub (XU, Mean2D)")
                 .Add("Normalized = Div (Deviation, StdDev)")
```
Comment (pengwa) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2098]: >Normalized [](http://example.com/codeflow?start=22&length=10)

This is a similar issue with the previous comment. 

Normalized: U, NormalizedV: V, Scale2D: V

When U = float, V = half, I think the following computation is not safe. 
Normalized in U --> Normalized in V --> Scaled in V

In our kernel registration, type U > V, U > T, unless we changed that,  otherwise we can reorder the compute to fix this problem.

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 3]
```diff
@@ -1982,22 +1982,25 @@ void RegisterContribSchemas() {
...
+      .Output(0, "Y", "Output data tensor.", "V")
       .Output(1, "Mean", "Saved mean used during training to speed up gradient computation", "U", OpSchema::Optional)
       .Output(2, "InvStdDev", "Saved inverse standard deviation used during training to speed up gradient computation.", "U", OpSchema::Optional)
       .TypeConstraint(
           "T",
           {"tensor(float16)", "tensor(float)", "tensor(double)", "tensor(bfloat16)"},
-          "Constrain input types and output Y type to float tensors.")
+          "Constrain input X type to float tensors.")
       .TypeConstraint(
           "U",
```
Comment (pengwa) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1995]: Not related to your change. 'U' did not allow double, how could register the kernel like this?  

      class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 1, double_double_double, LayerNormalization);

File: orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc [hunk 4]
```diff
@@ -11,34 +11,25 @@ namespace contrib {
...
-  ONNX_OPERATOR_TYPED_KERNEL_EX(                                  \
-      InvertibleLayerNormalizationGrad,                           \
-      kMSDomain,                                                  \
-      1,                                                          \
-      T,                                                          \
-      kCpuExecutionProvider,                                      \
-      KernelDefBuilder()                                          \
-          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
-      InvertibleLayerNormGrad<T>);
+#define REGISTER_KERNEL_TYPED(T)                                                                          \
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc:14]: Is this the new clang-formatting?

File: onnxruntime/contrib_ops/cuda/layer_norm_impl.cu [hunk 5]
```diff
@@ -326,18 +326,18 @@ __global__ void cuApplyLayerNorm(
     U mu, sigma2;
     cuWelfordMuSigma2<T, U, simplified>(vals, n1, n2, i1, mu, sigma2, buf);
     const T* lvals = vals + i1 * n2;
-    T* ovals = output_vals + i1 * n2;
+    V* ovals = output_vals + i1 * n2;
     U c_inv_std_dev = rsqrt(sigma2 + epsilon);
```
Comment (pengwa) [inline onnxruntime/contrib_ops/cuda/layer_norm_impl.cu:330]: Seems 'U' is used to do the safe operation. it is specified as attribute of LayerNorm op, named 'stash_type'.

Comment (pengwa) [reply onnxruntime/contrib_ops/cuda/layer_norm_impl.cu:330]: So U type is used for calculation, then cast to V finally into ovals[i].

File: orttraining/orttraining/training_ops/cuda/nn/layer_norm.h [hunk 6]
```diff
@@ -4,7 +4,7 @@
 
 namespace onnxruntime {
 namespace cuda {
-template <typename T, typename U>
+template <typename T, typename U, typename V>
 class LayerNorm final : public CudaKernel {
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cuda/nn/layer_norm.h:8]: Is this class still usable?

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 7]
```diff
@@ -2216,40 +2220,53 @@ Example 4:
...
 
             // Requirements/assumptions:
-            // Inputs Y_grad and X are of shape [d[0], ..., d[axis-1], d[axis], ..., d[rank-1]] and type T
-            // Input scale is of shape [d[axis], ..., d[rank-1]] and type U
-            // Inputs mean and inv_std_dev are of shape [d[0], ..., d[axis-1], 1, ..., 1] (same rank as X)
-            // and type U.
+            // Inputs Y_grad and X are of shape [d[0], ..., d[axis-1], d[axis], ..., d[rank-1]]
+            // Input scale is of shape [d[axis], ..., d[rank-1]]
+            // Inputs mean and inv_std_dev are of shape [d[0], ..., d[axis-1], 1, ..., 1] (same rank as X).
+            // Cast to type U for calculation for batter precision.
```
Comment (pengwa) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2244]: "better"

File: orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc [hunk 8]
```diff
@@ -11,34 +11,25 @@ namespace contrib {
...
+                                    .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())                \
+                                    .TypeConstraint("U", DataTypeImpl::GetTensorType<T>())                \
+                                    .TypeConstraint("V", DataTypeImpl::GetTensorType<T>()),               \
+                                LayerNormGrad<T, true>);                                                  \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(InvertibleLayerNormalizationGrad, kMSDomain, 1, T, kCpuExecutionProvider, \
+                                KernelDefBuilder()                                                        \
+                                    .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())                \
+                                    .TypeConstraint("U", DataTypeImpl::GetTensorType<T>())                \
+                                    .TypeConstraint("V", DataTypeImpl::GetTensorType<T>()),               \
+                                InvertibleLayerNormGrad<T>);
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc:33]: We missed to update the new types in this files. 

    template <typename T, bool simplified>
    LayerNormGrad<T, simplified>::LayerNormGrad(const OpKernelInfo& op_kernel_info)

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 9]
```diff
@@ -62,6 +62,19 @@ X --> ReduceMean --> Sub -->  Pow --> ReduceMean --> Add --> Sqrt --> Div --> Mu
...
+When using Apex O2, a Cast node may be inserted between Div and Mul, Layer Normalization will also handle the case below:
++---------------------+
+|                     |
+|                     v
+X --> ReduceMean --> Sub --> Pow --> ReduceMean --> Add --> Sqrt --> Div --> Cast --> Mul --> Add
+                      |                                               ^
+                      |                                               |
+                      +-----------------------------------------------+
+
+Since LayerNormalization supports input and scale/bias have different data type, and during the kernel execution,
```
Comment (pengwa) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:75]: nit: "have different data type" -> 'in different data types'

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 10]
```diff
@@ -62,6 +62,19 @@ X --> ReduceMean --> Sub -->  Pow --> ReduceMean --> Add --> Sqrt --> Div --> Mu
...
+data are casted to float/double to calculate for precision, so if there is a Cast before ReduceMean, we can also
+remove this Cast.
```
Comment (pengwa) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:77]: Can we add the optional "Cast" node before ReduceMean in the commented graph?

Comment (pengwa) [reply onnxruntime/core/optimizer/layer_norm_fusion.cc:77]: And need to clarify which ReduceMean among the two.

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 11]
```diff
@@ -117,6 +130,47 @@ Status LayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
...
+        continue;
+      }
+    }
+
+    const Node* p_reduce_mean_input_node = graph_utils::GetInputNode(reduce_mean_node, 0);
+    bool has_leading_cast = false;
+    if (p_reduce_mean_input_node) {
+      Node& reduce_mean_input_node = *graph.GetNode(p_reduce_mean_input_node->Index());
+      // If input to Pow is a Cast, and the Cast has same consumer count as subCnt + 1
+      if (graph_utils::IsSupportedOptypeVersionAndDomain(reduce_mean_input_node, "Cast", {9, 13}) &&
```
Comment (pengwa) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:166]: don't need check Pow op or not?

File: onnxruntime/core/optimizer/layer_norm_fusion.cc (hunk 9 above)
Comment (pengwa) [reply onnxruntime/core/optimizer/layer_norm_fusion.cc:75]: Can you help name this new pattern differently, for example "Apex O2", for the blocks that are specific for this pattern, we add comment 

"Apex O2 pattern specific match starts"
...
"Apex O2 pattern specific  match ends"

Once the change got merged, it's really hard to understand which piece of code is doing what pattern match....

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 12]
```diff
@@ -222,8 +276,15 @@ Status LayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
       }
     }
 
-    // div --> mul
-    Node& mul_node = *graph.GetNode(div_node.OutputNodesBegin()->Index());
+    // div --> mul or div --> cast --> mul
+    Node* next_node = graph.GetNode(div_node.OutputNodesBegin()->Index());
+    if (graph_utils::IsSupportedOptypeVersionAndDomain(*next_node, "Cast", {9, 13}) &&
```
Comment (pengwa) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:281]: I think this is also a "Apex O2 pattern specific match " block.

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 13]
```diff
@@ -302,7 +363,10 @@ Status LayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
     if (!same_dim)
       continue;
 
-    const std::array layer_norm_input_defs{reduce_mean_node.MutableInputDefs()[0], scale, bias};
+    const std::array layer_norm_input_defs{has_leading_cast
+                                               ? graph.GetNode(p_reduce_mean_input_node->Index())->MutableInputDefs()[0]
```
Comment (pengwa) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:367]: the intent looks a bit weird.

**Comment by pengwa** [comment]
Sorry for the late response, this PR has been there for more than two months. While I do think we need such a change, especially about the different input types support for LayerNorm/LayerNormGrad, considering we have prove the perf gains in one of customer models. 

The major comment is, for LayerNorm kernels, we need make sure all computation are using a promoted type to store the intermediate values, or following exiting approach, use the U type as intermediate variable types (as we did at many places in LayerNormGrad). And as said, LayerNormGrad current implementation is safe, because U is used.

**Comment by pengwa** [comment]
> 

nit: I am curious how this compiles without implement all template type we defined in the op schema.

---
Refers to: onnxruntime/contrib_ops/cpu/layer_norm.cc:38 in 726116e. [](commit_id = 726116eebc0ddea662ae87b8760fb39765d6cb0b, deletion_comment = False)

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 1 above)
Comment (iK1D) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1987]: U is the type for calculation internal, normally float (double for double T and V). Previously we require input, weight and bias are all T, T can be float or half, now we can support T for input, V for weight and bias, they can be different, and are are smaller than U in size. The output type is same as weight and bias, which is V. During the calculation, we will cast data to U for calculation internally, so we will not loss precision. So for kernel registration, we don't support U as half. Actually this kernel is copied from apex, but from an older version which have T and U only. Apex already updated their code to support T, V and U, I just followed the same way to move the new version of code from Apex to here... The code is here: https://github.com/NVIDIA/apex/blob/master/csrc/layer_norm_cuda_kernel.cu

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 3 above)
Comment (iK1D) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1995]: Right. Should double instead of bfloat16.

File: orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc (hunk 4 above)
Comment (iK1D) [reply orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc:14]: Using the .clang-format file and the VSCode's "Format document/selection", the code will be formatted like this. It will try to use the 120 columns as possible as it can instead of starting a new line for each argument or function caller. So the answer is yes, or maybe the old code just not the correct clang-format in the .clang-format file.

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 1 above)
Comment (pengwa) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1987]: I see, thanks for the explanation!

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 2 above)
Comment (iK1D) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2098]: I think this is same logic as the sub-graph we want to fuse. The Cast Op is added right before the final Mul and Add from float to half. As said, the internal logic is on U for precision, then when it goes to the end, to produce V output, we cast it back to V and then do final calculation. One way is to cast both scale and bias to U, calculate and then cast back to V for output, but this is not necessary as we add more Cast Ops.

File: orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc (hunk 8 above)
Comment (iK1D) [reply orttraining/orttraining/training_ops/cpu/nn/layer_norm.cc:33]: For cpu kernel, all inputs and outputs are in T. The class is for kernel implementation, only T is needed. During kernel registry, we will map all T, U, V to the given type we want to register.

File: onnxruntime/core/optimizer/layer_norm_fusion.cc (hunk 11 above)
Comment (iK1D) [reply onnxruntime/core/optimizer/layer_norm_fusion.cc:166]: Wrong comment... Nothing to do with Pow, should not be Pow here, but the first ReduceMean.

File: orttraining/orttraining/training_ops/cuda/nn/layer_norm.h (hunk 6 above)
Comment (iK1D) [reply orttraining/orttraining/training_ops/cuda/nn/layer_norm.h]: Seems not. Removed it.

File: onnxruntime/core/optimizer/layer_norm_fusion.cc (hunk 13 above)
Comment (iK1D) [reply onnxruntime/core/optimizer/layer_norm_fusion.cc:384]: Formatted by VScode...

**Review by pengwa — APPROVED** [review]
LGTM. Pls fix the build issues.

**Review by pengwa — DISMISSED** [review]

File: onnxruntime/contrib_ops/cpu/layer_norm.cc [hunk 14]
```diff
@@ -12,25 +12,19 @@
...
+                                    .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())              \
+                                    .TypeConstraint("U", DataTypeImpl::GetTensorType<T>())              \
+                                    .TypeConstraint("V", DataTypeImpl::GetTensorType<T>()),             \
+                                LayerNorm<T, false>);                                                   \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(SimplifiedLayerNormalization, kOnnxDomain, 1, T, kCpuExecutionProvider, \
+                                KernelDefBuilder()                                                      \
+                                    .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())              \
+                                    .TypeConstraint("U", DataTypeImpl::GetTensorType<T>())              \
+                                    .TypeConstraint("V", DataTypeImpl::GetTensorType<T>()),             \
+                                LayerNorm<T, true>);
```
Comment (skottmckay) [inline onnxruntime/contrib_ops/cpu/layer_norm.cc:28]: We need the hash to be stable for backwards compatibility. As we don't have a good way to handle completely new constraints being added (we can handle the types in the list changing) without an opset change (contrib ops are all opset 1), we have a list of hardcoded exceptions in UpdateHashForBackwardsCompatibility in \onnxruntime\core\framework\kernel_def_hash_helpers.cc. 

Please add a mapping from the original hash to the new one for the 4 CPU contrib ops there (and note exactly which registration, including type, each hash is for in the comment).

Comment (iK1D) [reply onnxruntime/contrib_ops/cpu/layer_norm.cc:28]: Added. I cann't tell which ones are for float or double, so I assume the first ones in contrib.cpu.json is for float. And we don't need this for training ops, right?

**Review by skottmckay — DISMISSED** [review]

**Review by skottmckay — APPROVED** [review]
